### ISSUE TITLE
Any example of pull to refresh working with this library?

### ISSUE THREAD

**Issue by ColtonIdle** [KevinnZou/compose-webview#7]
Pull to refresh with accompanist webview seems to be a fairly popular request according to stackoverflow: https://stackoverflow.com/questions/75751943/implement-pullrefresh-on-top-of-webview-accompanist-in-compose

Looks like it wasn't really working with accompanist but it also seems like according to ben trengrove that it should be possible
https://github.com/google/accompanist/issues/1670

i wonder if we can have some docs or something around successfully adding pull to refresh with this library. I've had some luck implementing it like so (material3 pull to refresh)

```
            webviewtestTheme {
                val refreshState = rememberPullToRefreshState()

                val state = rememberWebViewState("https://cnn.com")

                Box(Modifier.nestedScroll(refreshState.nestedScrollConnection)) {
                    WebView(state, modifier = Modifier.verticalScroll(rememberScrollState()))
                    PullToRefreshContainer(
                        modifier = Modifier.align(Alignment.TopCenter),
                        state = refreshState,
                    )
                }
            }

```
but... scrolling is now a bit wonky (i think because of the modifier i set on the webview) **and** it seems to cause issues if a webpage is short (which is also mentioned in the stackoverflow answers).

**Comment by KevinnZou** [comment]
@ColtonIdle Thank you for sharing your feedback! Your suggestion is excellent. My first idea is to use a scrollable container to wrap the WebView. You can then use this container to interact with pull-to-refresh. Additionally, you'll need to handle the nested scroll of the WebView and the container with a custom nested scroll connection. You can refer to this [library](https://github.com/KevinnZou/compose-nestedscroll-webview) for a sample of nested scrolling. I will try to provide a sample for this issue, but it may take some time as I am currently busy with other matters. If you work it out, you are welcome to submit a PR. Once again, thank you for your valuable suggestion!

**Comment by KevinnZou** [comment]
@ColtonIdle I tried a simple approach that seems to work. Check out the [branch](https://github.com/KevinnZou/compose-webview/tree/feature/pull_to_refresh) for details. For pull-to-refresh functionality, I used this [library](https://github.com/KevinnZou/compose-pulltorefresh). Please let me know if you have any questions. Thanks!

**Comment by ColtonIdle** [comment]
Thanks! The functionality (and issue) that my solution above had are still present in your sample branch. Basically it works on like 90% of websites... but some websites are "short" (like login pages especially) and webview acts differently when its a scrollable container, and won't actually fillMaxSize.

So you end up with something that looks broken like 

val initialUrl = "https://boomboomarmy.com/password"
![CleanShot 2024-04-03 at 18 28 18@2x](https://github.com/KevinnZou/compose-webview/assets/31751141/e197b030-4c5f-4ea6-8528-92bb24aa4207)

You can see people complaining about it here: "For web pages that are longer than the screen it works like a charm. But for shorter pages I couldn't get them to work in fullscreen yet."

https://stackoverflow.com/questions/75751943/implement-pullrefresh-on-top-of-webview-accompanist-in-compose

So basically. it (seems) impossible to have a full page webview when using pull to refresh. which makes login pages unusable

**Comment by KevinnZou** [comment]
@ColtonIdle Thanks for your information! I was able to reproduce the issue as well. You are right, it appears to be a problem with Jetpack Compose and we are unable to solve it on our end. I recommend submitting an issue to Jetpack Compose and waiting for their solution.

**Comment by ColtonIdle** [comment]
Yeah, I did get in touch with Ben from accompanist. he said something along the lines of ~"This was something we had and then removed at some point because it broke things like this with nested scrolling" and basically made it seem like this would have to be solved by forking webview. the google issue about it was marked as wontdo. I will try to hack a fork of webview together that can support this. if you have any tips on where to start... let me know! thanks

**Comment by KevinnZou** [comment]
@ColtonIdle Certainly, feel free to submit a pull request directly to this repository. Thank you!

**Comment by ColtonIdle** [comment]
Thanks. I'll try to take a look this weekend. If you have any idea what could be the cause of it (not sure how familiar you are with compose webview internals), let me know. I'm not sure where to start 😅

**Comment by ColtonIdle** [comment]
😂 just noticed you created one of the issues here too https://github.com/google/accompanist/pull/1684

i will try by commenting out the isNestedScrollingEnabled and maybe that'll fix it? i dont need nested scrolling. im using an app thats essentially a webview so i need the whole thing to take up the whole page

**Comment by KevinnZou** [comment]
> 😂 just noticed you created one of the issues here too [google/accompanist#1684](https://github.com/google/accompanist/pull/1684)
> 
> i will try by commenting out the isNestedScrollingEnabled and maybe that'll fix it? i dont need nested scrolling. im using an app thats essentially a webview so i need the whole thing to take up the whole page

If you want to use pull to refresh, I think you cannot simply comment out the isNestedScrollingEnabled. Otherwise, outer pull to refresh container will not receive the event.

**Comment by ColtonIdle** [comment]
Yeah, I tried hacking something together over the span of a few hours today. still no luck with a working solution 😭 

such an interesting problem! didn't think adding pull to refresh was going to be this difficult with compose!

**Comment by ycngmn** [comment]
Hello @KevinnZou, 
Can you please take a look at this approach for pull to refresh and maybe try to adapt this into the library ? 

```kotlin
fun WebViewScreen(
    viewModel: WebViewViewModel,
    modifier: Modifier = Modifier
) {
    val refreshScope = rememberCoroutineScope()
    var refreshing by remember { mutableStateOf(false) }
    AndroidView(
        modifier = modifier
            .fillMaxSize(),
        factory = { context ->
            SwipeRefreshLayout(context).apply {
                viewModel.webView.setOnScrollChangeListener { view, scrollX, scrollY, oldScrollX, oldScrollY ->
                    isEnabled = scrollY == 0
                }
                setOnRefreshListener {
                    refreshScope.launch {
                        refreshing = true
                        viewModel.webView.reload()
                        delay(1500)
                        refreshing = false
                    }

                }
                addView(viewModel.webView)
            }
        },
        update = { view ->
            view.isRefreshing = refreshing
            viewModel.webView.loadUrl(viewModel.url)
        }
    )
}
```

From : [Stackoverflow](https://stackoverflow.com/a/78179704/11724248)
